crates/runtime/src/instance/allocator/pooling/index_allocator.rs

```diff
@@ -0,0 +1,408 @@
+            },
+        }
+    }
+
+    /// Are any slots left, or is this allocator empty?
+    pub(crate) fn is_empty(&self) -> bool {
+        match self {
+            &PoolingAllocationState::NextAvailable(ref free_list)
+            | &PoolingAllocationState::Random(ref free_list) => free_list.is_empty(),
+            &PoolingAllocationState::ReuseAffinity { ref free_list, .. } => free_list.is_empty(),
+        }
+    }
+
+    /// Internal: remove a slot-index from the global free list.
+    fn remove_free_list_item(
+        slot_state: &mut Vec<SlotState>,
+        free_list: &mut Vec<usize>,
+        index: usize,
+    ) {
+        let free_list_index = slot_state[index].free_list_index();
+        assert_eq!(index, free_list.swap_remove(free_list_index));
+        if free_list_index < free_list.len() {
+            let replaced = free_list[free_list_index];
+            slot_state[replaced].update_free_list_index(free_list_index);
+        }
+    }
+
+    /// Internal: remove a slot-index from a per-module free list.
+    fn remove_module_free_list_item(
+        slot_state: &mut Vec<SlotState>,
+        per_module: &mut HashMap<CompiledModuleId, Vec<usize>>,
+        id: CompiledModuleId,
+        index: usize,
+    ) {
+        let per_module_list = per_module.get_mut(&id).unwrap();
+        let per_module_index = slot_state[index].per_module_index();
+        assert_eq!(index, per_module_list.swap_remove(per_module_index));
+        if per_module_index < per_module_list.len() {
+            let replaced = per_module_list[per_module_index];
+            slot_state[replaced].update_per_module_index(per_module_index);
+        }
+        if per_module_list.is_empty() {
+            per_module.remove(&id);
+        }
+    }
+
+    /// Allocate a new slot.
+    pub(crate) fn alloc(&mut self, id: Option<CompiledModuleId>) -> usize {
+        match self {
+            &mut PoolingAllocationState::NextAvailable(ref mut free_list) => {
+                debug_assert!(free_list.len() > 0);
+                free_list.pop().unwrap()
+            }
+            &mut PoolingAllocationState::Random(ref mut free_list) => {
+                debug_assert!(free_list.len() > 0);
+                let id = rand::thread_rng().gen_range(0..free_list.len());
+                free_list.swap_remove(id)
+            }
+            &mut PoolingAllocationState::ReuseAffinity {
+                ref mut free_list,
+                ref mut per_module,
+                ref mut slot_state,
+                ..
+            } => {
+                if let Some(this_module) = id.and_then(|id| per_module.get_mut(&id)) {
+                    // There is a freelist of slots with affinity for
+                    // the requested module-ID. Pick the last one; any
+                    // will do, no need for randomness here.
+                    assert!(!this_module.is_empty());
+                    let new_id = this_module.pop().expect("List should never be empty");
```

Comment:
It would be nice to have a newtype for slot indices (`SlotId`?) too.

`new_id` threw me for a little bit of a loop, and made me double check that this wasn't a `CompiledModuleId` and was actually a slot index, which is what I expected it to be, but just the naming made me doubt myself enough that I had to double check.

Reply:
and/or we could `s/new_id/slot_index/` here

Reply:
I newtyped All The Things (`SlotId`, `GlobalFreeListIndex`, `PerModuleFreeListIndex`); hopefully this is better!